src/unix/vsock.c

```diff
@@ -0,0 +1,990 @@
+                success = ringbuf_write(s->incoming, &child, sizeof(child));
+                if (success)
+                    vsock_add_connection(conn);
+            } else {
+                success = false;
+            }
+            vsock_unlock(s);
+            if (!success)
+                apply(child->sock.f.close, 0, io_completion_ignore);
+        } else {
+            success = false;
+        }
+        blockq_wake_one(s->sock.rxbq);
+        fdesc_notify_events(&s->sock.f);
+    } else {
+        success = false;
+    }
+    spin_unlock(&bound->lock);
+    refcount_release(&bound->refc);
+    return success;
+}
+
+vsock_connection vsock_connect_complete(struct vsock_conn_id *conn_id, boolean success)
+{
+    vsock_connection conn = vsock_get_connection(conn_id);
+    if (!conn)
+        return conn;
+    vsock_conn_lock(conn);
+    vsock s = conn->vsock;
+    if (s) {
+        vsock_lock(s);
+        if (success) {
+            s->state = VSOCK_STATE_CONNECTED;
+        } else {
+            s->so_error = ECONNREFUSED;
+            s->state = VSOCK_STATE_BOUND;
+        }
+        vsock_unlock(s);
+        blockq_wake_one(s->sock.txbq);
+        notify_dispatch(s->sock.f.ns, vsock_events_connlocked(s));
+    }
+    if (success)
+        return conn;
+    vsock_remove_connection(conn);
+    vsock_conn_unlock(conn);
+    vsock_conn_release(conn);
+    return 0;
+}
+
+vsock_connection vsock_get_conn(struct vsock_conn_id *conn_id)
+{
+    vsock_connection conn = vsock_get_connection(conn_id);
+    if (conn)
+        vsock_conn_lock(conn);
+    return conn;
+}
+
+vsock_connection vsock_rx(struct vsock_conn_id *conn_id, void *data, u64 len)
+{
+    vsock_connection conn = vsock_get_connection(conn_id);
+    if (!conn)
+        return conn;
+    vsock_conn_lock(conn);
+    vsock s = conn->vsock;
+    if (s) {
+        vsock_lock(s);
+        if (s->flags & VSOCK_SHUTDOWN_RX)
+            len = 0;
+        boolean notify = (ringbuf_length(s->incoming) == 0) && (len > 0);
+        if (len > 0) {
+            struct vsock_rxbuf rxbuf = {
+                .data = data,
+                .offset = 0,
+                .len = len,
+            };
+            if (!ringbuf_write(s->incoming, &rxbuf, sizeof(rxbuf))) {
+                /* couldn't enqueue incoming data: disable data reception on this connection */
+                virtio_sock_free_rxbuf(vsock_priv.transport, data);
+                s->flags |= VSOCK_SHUTDOWN_RX;
+                virtio_sock_shutdown(conn, VSOCK_SHUTDOWN_RX);
+                notify = false;
+            }
+        } else {
+            virtio_sock_free_rxbuf(vsock_priv.transport, data);
+        }
+        vsock_unlock(s);
+        blockq_wake_one(s->sock.rxbq);
+        if (notify)
+            notify_dispatch(s->sock.f.ns, vsock_events_connlocked(s));
+    }
+    return conn;
+}
+
+void vsock_buf_space_notify(vsock_connection conn, u64 buf_space)
+{
+    vsock s = conn->vsock;
+    blockq_wake_one(s->sock.txbq);
```

Comment:
I got a few crashes here with a page fault that looked like txbq was null. As I turned on some of the vsock debugging, the problem went away and I haven't been able to reproduce it even after turning it back off. It's possible there was just a stale object file, but the crash happened after the "Hello from Nanos" message was delivered (following this article https://nanovms.com/dev/tutorials/what-is-vsock-why-use-with-unikernels), so I initially thought there was perhaps a race when cleaning up the socket. Still, I'm not sure it's a real problem and I unfortunately didn't capture the backtrace.

Reply:
Although I have been unable to reproduce any crash here, I think I know what may have happened: the VIRTIO_VSOCK_OP_RST packet sent by the host in response to the VIRTIO_VSOCK_OP_SHUTDOWN packet sent by the guest (when the socket is closed by the guest program) was processed by the virtio driver just before terminating the VM, and so `conn->vsock'  was NULL, leading to a null pointer dereference.
I added a check for a non-null pointer, just like it's done in other similar functions called by the transport layer when processing a received packet.